### DISCUSSION TITLE
TODO: @babel/preset-env & browserlist specification / 规定浏览器目标版本

### DISCUSSION THREAD

**Discussion by tonywu6** [antvis/X6#2212] [Ideas]
tl;dr 

⭐️ **收集 IE 11 的使用场景以及遇到的缺陷（涉及的功能及代码示例），收集和浏览器支持相关的看法和建议** ⭐️
⭐️ **We are collecting use cases in IE 11 (and other legacy browsers), issues you've encountered, and your opinions/suggestions on enabling better cross-browser compatibility.** ⭐️

---

目前我们在代码打包的过程中仅仅使用 `tsc` 将 JavaScript 的代码格式转译成我们期望支持的最低 ECMAScript 版本（目前是 ES5），而没有使用如 Babel 的工具将其进一步转化成跨浏览器支持的代码。这意味着我们没有办法明确定义 X6 支持哪一些浏览器，也没有办法在文档中提供这一信息。

转译的缺失导致了相当数量的浏览器兼容 issue （大部分来自 IE）

- #2203
- #1001
- #989

我们在 v2.0 版本中，极有可能会在 rollup 打包过程中引入 `@babel/preset-env` 来进行跨浏览器支持，并且使用 browserslist 来限定我们支持的浏览器版本。

可以讨论的两个问题是：

- 我们要不要将 ECMAScript target 提到 ES6 (ES2015)
- 我们要不要支持 IE 11

---

Currently, our build process only transforms our scripts' JavaScript syntax to the minimum ECMAScript version that we wish to support (currently ES5) using `tsc`. We do not transform our code using cross-browser transpiling tools such as Babel. This means we cannot determine or document precisely which browsers we support.

A bunch of issues were opened because of this lack of tooling (unsurprisingly most of them were about Internet Explorer):

- #2203
- #1001
- #989

As part of out v2.0 rewrite, we are likely to introduce Babel into our rollup workflow, allowing us to generate code that is guaranteed to work in a defined list of browser versions via [`@babel/preset-env`](https://babeljs.io/docs/en/babel-preset-env).

The following questions are then to be resolved:

- Should we move to a newer ECMAScript version, such as ES6 (ES2015)?
- Should we support IE 11 (via polyfill)?

---

EDIT: add tl;dr

**Comment by tonywu6** [comment]
我的看法：

- 我们要不要将 ECMAScript target 提到 ES6 (ES2015), 或者一个更新的版本
  - 要。ES2015 已经在[绝大部分浏览器中支持](https://caniuse.com/?search=ES6). 停留在 ES5 基本没有好处。
- 我们要不要支持 IE 11
  - [**不要。**](https://blog.logrocket.com/should-you-drop-support-for-internet-explorer-11/)去掉 IE 11 意味着我们会获得更小的 bundle 以及潜在更良好的性能，因为代码毋需使用 polyfill 转写，或是转写成更原始的 JavaScript 语法，而可以使用更新的 JS 特性。
  - IE 已经在很长一段时间处于被汰换的阶段了：Windows 11 已经不提供 IE 11, 大部分网站也开始去掉对它的支持，微软将在 [2022 年 6 月 15 日，也就是**这个评论的 5 天后**，正式停止维护 IE.](https://support.microsoft.com/en-us/windows/internet-explorer-downloads-d49e1f0d-571c-9a7b-d97e-be248806ca70)
  - 若是开发者的确需要 IE 支持，仍然可以选择将 X6 作为 ESM/CJS 模块导入自己的项目中，然后自己配置 Webpack 或者其它工具，将最终的源码自行转译成支持 IE 的版本。

---

My opinions:

- Should we move to a newer ECMAScript version, such as ES6 (ES2015), or an even newer version?
  - Very much yes, as ES2015 has a [global support of ~96%](https://caniuse.com/?search=ES6). There is virtually no benefit in staying with ES5.
- Should we still support IE 11 (via polyfill)?
  - [Personally **no.**](https://blog.logrocket.com/should-you-drop-support-for-internet-explorer-11/) Dropping support for IE means we will have potentially smaller bundles and better performance, because higher level browser APIs no longer have to be polyfilled or transpiled to a simpler version. 
  - Internet Explorer is going away: Windows 11 no longer provides it, [major websites are dropping its support too](https://www.zdnet.com/article/get-ready-to-switch-aws-is-dropping-support-for-ie11/), and Microsoft is [formally ending its support on **June 15, 2022, 5 days after this comment was posted.**](https://support.microsoft.com/en-us/windows/internet-explorer-downloads-d49e1f0d-571c-9a7b-d97e-be248806ca70)
  - Dropping IE support doesn't necessarily mean X6 will be entirely unusable in an IE environment. Developers can still use X6 in their projects as an ESM/CJS module, and then create their build workflow to transpile their final bundle to an IE-supporting version.

**Comment by NewByVector** [comment]
非常赞同 @tonywu6  的观点，不过第一个要问题要考虑到 umd 包能否在旧版本的 chrome 上能否跑起来。
